server/routerlicious/packages/services-core/src/runWithRetry.ts

```diff
@@ -73,3 +76,80 @@ export async function runWithRetry<T>(
+ * @param  {number} maxRetries - maximum retries after which error is thrown. Retry infinitely if set to -1
+ * @param  {number} retryAfterMs - interval factor to wait before retrying. Param to calculateIntervalMs
+ * @param  {(error)=>boolean} shouldRetry? - function that takes error and decides whether to retry on it
+ * @param  {(error, numRetries, retryAfterInterval)=>number} calculateIntervalMs
+ * function which calculates interval to wait before retrying based on error, retryAfterMs and retries so far
+ * @param  {(error)=>void} onErrorFn? - function allowing caller to define custom logic to run on error e.g. custom logs
+ */
+ export async function requestWithRetry<T>(
+    request: () => Promise<T>,
+    callName: string,
+    maxRetries: number,
+    retryAfterMs: number,
+    shouldRetry?: (error) => boolean,
+    calculateIntervalMs = (error, numRetries, retryAfterInterval) => retryAfterInterval * 2 ** numRetries,
+    onErrorFn?: (error) => void,
+): Promise<T> {
+    let result: T;
+    let retryCount = 0;
+    let success = false;
+    do  {
+        try {
+            result = await request();
+            success = true;
+            if (retryCount >= 1) {
+                Lumberjack.info(`Succeeded in executing ${callName} with ${retryCount} retries`);
+            }
+        } catch (error) {
+            Lumberjack.error(`Error running ${callName}: retryCount ${retryCount}`, undefined, error);
+            if (onErrorFn !== undefined) {
+                onErrorFn(error);
+            }
+            if (shouldRetry !== undefined && shouldRetry(error) === false)
+            {
+                Lumberjack.error(`Should not retry ${callName} for the current error, rejecting`, undefined, error);
+                return Promise.reject(error);
+            }
+            // if maxRetries is -1, we retry indefinitely
+            // unless shouldRetry returns false at some point.
+            if (maxRetries !== -1 && retryCount >= maxRetries) {
+                Lumberjack.error(`Error after retrying ${retryCount} times, rejecting`, undefined, error);
+                // Needs to be a full rejection here
+                return Promise.reject(error);
+            }
+
+            const intervalMs = calculateIntervalMs(error, retryCount, retryAfterMs);
+            await delay(intervalMs);
+            retryCount++;
+        }
+    } while (!success);
+
+    return result;
+}
+
+/**
+ * Helper function to decide when or not to retry a {@link NetworkError}.
+ * Can be used with {@link runWithRetry} and {@link requestWithRetry}.
+ * @param  {any} error - the error parameter to be inspected when deciding whether to retry or not.
+ */
+export function shouldRetryNetworkError(error: any): boolean {
+    if (error instanceof Error && error?.name === "NetworkError") {
+        const networkError = error as NetworkError;
+        return !networkError.isFatal && networkError.canRetry !== undefined && networkError.canRetry;
```

Comment:
nit: `networkError.canRetry !== undefined && networkError.canRetry` is redundant and could just be `networkError.canRetry` or `networkError.canRetry === true`

Reply:
I had it done that way because if I use
```
 return !networkError.isFatal && networkError.canRetry;
```
the return type would become `boolean | undefined`. Let me see if I can do it differently but still keeping the return type as `boolean`

Reply:
I think the `canRetry === true` would keep it as a boolean

Reply:
You are correct! I made the change and I'm testing it out now.